### PR TITLE
feat: exclude_dir for excluding directories within vaults

### PR DESCRIPTION
**Pull request by neuroconvergent**
# Excluding directories from workspaces (#816 )

This PR adds a config option `exclude_dir`, which accepts a table of path strings to exclude certain directories from the plug-in entirely. This follows a similar logic to the current implementation of `.gitignore` i.e. uses glob matching for path (removes all exclusions with ! from current gitignore implementation).

The config option has been tested to work as both a global option and as a workspace override. 

## PR Checklist

- [x] The PR contains a description of the changes
- [x] I read the [CONTRIBUTING.md] file
- [x] The `CHANGELOG.md` is updated
- [x] The changes are documented in the `README.md` file
- [x] The code complies with `make chores` (for style, lint, types, and tests)

### CODE REVIEW

File: lua/obsidian/exclusions.lua [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+local function get_vault_relative(path)
+  if not Obsidian or not Obsidian.dir then
+    return nil
+  end
+
+  local vault_root = vim.fs.normalize(tostring(Obsidian.dir))
+  path = vim.fs.normalize(path)
+
+  if vim.startswith(path, vault_root) then
```
Comment (neo451) [inline lua/obsidian/exclusions.lua]: looks like will produce false positives, either append a `/` in vault_root, or use `Path.is_parent_of`

File: lua/obsidian/search/opts.lua [hunk 2]
```diff
@@ -74,6 +74,12 @@ M._prepare = function(opts, additional_opts)
     add_exclude(search_opts, tostring(Obsidian.opts.templates.folder))
   end
 
+  if Obsidian.opts.exclude_dir and #Obsidian.opts.exclude_dir > 0 then
+    for _, dir in ipairs(Obsidian.opts.exclude_dir) do
+      add_exclude(search_opts, dir)
```
Comment (neo451) [inline lua/obsidian/search/opts.lua]: if the option supposed to be a list of globs or just subdir names? if former, the add_exclude will not work in many cases?

File: lua/obsidian/note.lua [hunk 3]
```diff
@@ -887,6 +888,10 @@ Note.should_save_frontmatter = function(self)
     end
   end
 
+  if exclusions.is_excluded(tostring(self.path)) then
```
Comment (neo451) [inline lua/obsidian/note.lua]: save frontmatter is too narrow for "completely making this plugin ignore", see `autocmds.lua`, there's a bufenter callback that attaches everything, keymaps, lsp and etc, just make that check fail if exclude seems more reasonable

**Review by neo451 — COMMENTED** [review]
see the code comments, plus:
1. there's an obsidian app counterpart that is internally called "userIgnoreFilters", so might be good to call the option "ignore_filters"
2. I am not too sure about why add a new option, instead of just use gitignore's patterns, make note attaching and ripgrep also ignore them, seems cleaner, I get the appeal to declare a readable list in config, but for now I prefer less option the better
3. need some tests, also fine if you prefer to not go down that rabbit hole, since we don't have proper test coverage in the search module, so just some simpler tests can do

**Comment by neo451** [comment]
thank you for the idea and work btw :)

**Comment by neuroconvergent** [comment]
My original idea was just to exclude entire sub-directories, but it might be better to rename the option to allow individual files. As you have already pointed out, it would fail with some glob patterns when looking so that needs to be robust.

The problem with just using gitignore is that there might be files which are to be committed to git but need to be ignored from the plug-in. For example, I use a program called `presenterm` for creating review slides but these markdown files don't work properly with some YAML headers used by Obsidian. So I want to exclude these files from Obsidian frontmatter or LSP while keeping them in my git tree. 

I can certainly add some tests, I hope a few simple tests with different kinds of options such as relative path, absolute path and glob patterns should suffice. I can't personally think of any edge cases that specifically need to be tested but do let me know if you have any in mind and I can add them.

File: lua/obsidian/search/opts.lua (hunk 2 above)
Comment (neuroconvergent) [reply lua/obsidian/search/opts.lua]: I did some testing and glob patterns do seem to work. Only exception seems to be a global path but that is the expected ripgrep behaviour.

Can you maybe give me an example where this would fail?

**Comment by neo451** [comment]
great work, sorry for the late response, some notes before I merge this:
1. address the style thing
2. I think this can be moved into a new `file` config module, like `file.ignore_filters`, and there can be a `file.trash` with #804 
3. for consistency, rename all the `exclude` `exclusion` in code to `ignore`, the module name and the function names
4. document in a new `docs/File.md` and in inline comments, that users should use simple gitignore style globs without modifiers, and rg compatibility is not guaranteed
5. in `autocmds.lua`, move the ignore check after `find_workspace` call, otherwise it is called for any markdown file, which is not efficient and outside the duty of this plugin.

**Comment by neuroconvergent** [comment]
These suggestions seem fair to me so I have made the changes you suggested.

**Comment by neo451** [comment]
thank you for the great work!
